### ISSUE TITLE
Get instance of Multiple editors in react

### ISSUE THREAD

**Issue by maneeshbhunwal123** [tinymce/tinymce-react#43]
I have two instances of react editor on my page. i want to insert image in them independently. I am creating my react widget like this

```
class CustomWisivigComponent extends React.Component {
    constructor(props) {
        super(props);
        this.onImageInsert = this.onImageInsert.bind(this);
        this.returnCmp = <Editor
            apiKey={"mypreciousKey"}
            init={{
                setup: (editor) => this.setState({editorRef: editor}),
                height: 700,
                plugins: 'image table link code',
                selector: "textarea",  // change this value according to your HTML
                default_link_target: "_blank"

            }}
        />
    }
//this is called from other components
    onImageInsert(propd) {
        //TODO fix the height and width.
        if(propd.fileType && propd.fileType == 'Image') {
            this.state.editorRef.insertContent('<img alt="Smiley face" height="${height}" width="${width}" src="${image_url}"/>'.replace("${image_url}",propd.AssetIdentifiers.callbackUrl).replace("${width}",propd.dimension.width).replace("${height}",propd.dimension.height));
        } else {
            this.state.editorRef.insertContent('<a href="${image_url}">document</a>'.replace("${image_url}",propd.AssetIdentifiers.callbackUrl));

        }
    }

    render() {
        return (
            <div>
                    {this.returnCmp}
            </div>
        );

    }
}

But this.state.editorRef.insertContent  will always insert image in the last wisiwig of the page instead of whichever is active. Am i missing anything?

**Comment by fyrkant** [comment]
Why put the editor in this.returnCmp property on the class? I've never seen it done like that before.

**Comment by bhunwal** [comment]
@fyrkant Thanks for your review, I can make it variable as well, but that wont solve the issue i am facing right? basically this component returns wysiwyg with some more custom functionality like image insert which is failing, can you please suggest something?

**Comment by fyrkant** [comment]
No I mean, why not just add it in the render directly? And you should also not have to define a selector in the init, the react wrapper does that automatically. Not sure if that is causing the problem you are having, though. How does the component where you are using this component look like?

**Comment by bhunwal** [comment]
I am using this component with [json-schema-form](https://github.com/mozilla-services/react-jsonschema-form). 
I will use the above widget to rendered directly. The only catch i am facing is, onImageInsert is always called upon the last wisiwig instead of the active one. there is no issue when only one  wisiwig is there.

**Comment by bhunwal** [comment]
@fyrkant breaking down the problem, is there any other way to insert image in react widget other than how I am doing?

**Comment by bhunwal** [comment]
can someone please help with this?

**Comment by fyrkant** [comment]
So you are saving references to the editors in state and then running insertContent on the editor reference in state? How do you set which is the active editor, when you focus it? 

Can you show your whole application or make a small app reproducing the same problem?

**Comment by bhunwal** [comment]
@fyrkant you got the problem exactly, that is the part i am struggling with, So the problem is in a page how will we associate id with a particular tinymice instance? hence need your help

**Comment by bhunwal** [comment]
@fyrkant In my understanding i necessarily know which editor is being currently used, since every editor has a button associated with it. when ever that button is clicked i expect image/object to be inserted in the editor being referenced in the state. does it make sense?

**Comment by bhunwal** [comment]
@fyrkant i created https://github.com/tinymce/tinymce-react/issues/46 with more info.

**Comment by fyrkant** [comment]
No reason to create a new issue, add the new information to this one instead.

**Comment by bhunwal** [comment]
I need to get editor reference since i am trying to implement wordpress like image insert feature, where image insert code will be owned by me and I will use the command 
`editorRef.insertContent('<a href="${image_url}">document</a>');`

But when I have multiple instances of tinymce i am not able to get editorRef correctly.
Here is my code
```
import React from 'react';
import { Editor } from '@tinymce/tinymce-react';

class CustomWisivigComponent extends React.Component {
    constructor(props) {
        super(props);
        this.putWisivigRef = this.putWisivigRef.bind(this);
        this.returnCmp = <Editor
            apiKey={"myKey"}
            init={{
                setup: (editor) => this.putWisivigRef(editor),
                height: 700,
                plugins: 'image table link code',
                selector: "textarea",  // change this value according to your HTML
                default_link_target: "_blank"

            }}
            initialValue={this.props.value}
            onEditorChange={(event) => {
                this.props.onChange(event)
            }}
        />
    }

    putWisivigRef(ref) {
        console.log("setting editorRef to "+ref.id + " in component with props="+JSON.stringify(this.props))
        this.setState({editorRef: ref})
    }

    onImageInsert(propd) {
        console.log(this.state.editorRef)
    }

    render() {

        return (
            <div>
                {this.returnCmp}
            </div>
        );

    }
}

export default CustomWisivigComponent;
```

And I am trying to render it like this
```
<CustomWisivigComponent id="1"/>
  <CustomWisivigComponent id="2"/>
```

The problem is when i execute the code putWisivigRef is called with two references but the context of this is missed and always the putWisivigRef of first component is called. And hence when I try to insert something in the first component, it gets inserted into second (since first one's reference gets overridden) and when I try to insert in 2nd it results into null pointer, since it was never called. I tried passing this from the setup: (editor) => this.putWisivigRef(editor) as well but that too did not work.

when we render the above form it logs

<img width="1309" alt="screen shot 2018-12-30 at 3 09 48 pm" src="https://user-images.githubusercontent.com/4371726/50545941-17052280-0c45-11e9-9146-6a6906e7302e.png">

but ideally it should have been

setting editorRef to tiny-react_10038262721546162755458 in component with props={"id":"1"}
setting editorRef to tiny-react_19726510231546162755460 in component with props={"id":"2"}

**Comment by fyrkant** [comment]
I really don't understand how you want the component to work even from looking at your code, how are you choosing which editor to insert an image into, the ref is just set in the local state?

**Comment by bhunwal** [comment]
My bad, if you could not understand.
Let me try explain again,
User may not be actually editing the content. I will have a insert media button on the top of every tinymce Please check the attached image. Now there a callback function in every insert image component. That component need to have a reference of tinymce to insert image in it.
<img width="1016" alt="screen shot 2019-01-08 at 5 23 29 pm" src="https://user-images.githubusercontent.com/4371726/50829352-406f2e00-136a-11e9-8eba-40fc5477a136.png">

**Comment by bhunwal** [comment]
@fyrkant do you need more info? please let me know, if this is still not clear, this is one last piece i am blocked on with tinymce integration

**Comment by bhunwal** [comment]
I have given all the code i have :). breaking down the problem if you just use the component twice in any react component you should be able to reproduce the problem. my point here is that when 
```setup: (editor) => this.putWisivigRef(editor),``` is called the component reference is lost. so i believe there is some gaucha in that function.

**Comment by bhunwal** [comment]
Would you be ok to have a call around this issue?

**Comment by fyrkant** [comment]
No I don't think that would help. I feel like you approach is unnecessarily complex, you could just give the editor an `id` and then make a button that inserts the image or whatever into the editor by taking the same `id` in as a prop and using `tinymce.get(id).insertContent(content)` or something like that?
